server/expression/in_subquery.go

```diff
@@ -0,0 +1,259 @@
+	"github.com/dolthub/go-mysql-server/sql/expression"
+	"github.com/dolthub/go-mysql-server/sql/plan"
+	"github.com/dolthub/go-mysql-server/sql/types"
+	vitess "github.com/dolthub/vitess/go/vt/sqlparser"
+
+	"github.com/dolthub/doltgresql/server/functions/framework"
+	pgtypes "github.com/dolthub/doltgresql/server/types"
+)
+
+// InSubquery represents a VALUE IN (SELECT ...) expression.
+type InSubquery struct {
+	leftExpr  sql.Expression
+	rightExpr *plan.Subquery
+
+	// These variables are used so that we can resolve the comparison functions once and reuse them as we iterate over rows.
+	// These are assigned in WithChildren, so refer there for more information.
+	leftLiteral   *Literal
+	rightLiterals []*Literal
+	compFuncs     []*framework.CompiledFunction
+}
+
+var _ vitess.Injectable = (*InSubquery)(nil)
+var _ sql.Expression = (*InSubquery)(nil)
+var _ expression.BinaryExpression = (*InSubquery)(nil)
+
+// NewInSubquery returns a new *InSubquery.
+func NewInSubquery() *InSubquery {
+	return &InSubquery{}
+}
+
+// Children implements the sql.Expression interface.
+func (in *InSubquery) Children() []sql.Expression {
+	return []sql.Expression{in.leftExpr, in.rightExpr}
+}
+
+var nilKey, _ = sql.HashOf(sql.NewRow(nil))
+
+// Eval implements the sql.Expression interface.
+func (in *InSubquery) Eval(ctx *sql.Context, row sql.Row) (any, error) {
+	if len(in.compFuncs) == 0 {
+		return nil, fmt.Errorf("%T: cannot Eval as it has not been fully resolved", in)
+	}
+
+	left, err := in.leftExpr.Eval(ctx, row)
+	if err != nil {
+		return nil, err
+	}
+
+	// The NULL handling for IN expressions is tricky. According to
+	// https://dev.mysql.com/doc/refman/8.0/en/comparison-operators.html#operator_in:
+	// To comply with the SQL standard, IN() returns NULL not only if the expression on the left hand side is NULL, but
+	// also if no match is found in the list and one of the expressions in the list is NULL.
+	// However, there's a strange edge case. NULL IN (empty list) return 0, not NULL.
+	leftNull := left == nil
+
+	if types.NumColumns(in.Left().Type()) != types.NumColumns(in.Right().Type()) {
+		return nil, sql.ErrInvalidOperandColumns.New(types.NumColumns(in.Left().Type()), types.NumColumns(in.Right().Type()))
+	}
+
+	typ := in.rightExpr.Type()
+	right := in.rightExpr
+
+	// TODO: does this work for all pg values?
+	values, err := right.HashMultiple(ctx, row)
+	if err != nil {
+		return nil, err
+	}
+
+	// NULL IN (list) returns NULL. NULL IN (empty list) returns 0
+	if leftNull {
+		if values.Size() == 0 {
+			return false, nil
+		}
+		return nil, nil
+	}
+
+	// convert left to right's type
+	nLeft, _, err := typ.Convert(left)
+	if err != nil {
+		return false, nil
+	}
+
+	key, err := sql.HashOf(sql.NewRow(nLeft))
+	if err != nil {
+		return nil, err
+	}
+
+	val, notFoundErr := values.Get(key)
+	if notFoundErr != nil {
+		if _, nilValNotFoundErr := values.Get(nilKey); nilValNotFoundErr == nil {
+			return nil, nil
+		}
+		return false, nil
+	}
```

Comment:
I may not be understanding what this is supposed to do, but I'm not finding any analogue to this in the documentation, so it seems like it's wrong. `values.Get` is comparing hashes, but Postgres seems to only be using the equality operator. So this would also lead to incorrect results for valid queries, since we're deviating from how Postgres is supposed to work.

Reply:
I added a comment. We're using the hash to quickly find a matching row, but it can return false positives, so we also check equality with the comparison operator.

If two values are considered equal by the comparison function but hash to different things, that will lead to false negatives. Not sure if this can happen or not, but we do need to be aware of it. I believe this is the point of the type conversion you commented on (although obviously that needs to work differently for postgres than just calling Convert. The hash comparisons speed things up by several orders of magnitude, so we can't just ditch it. For now I'm going to add a note and come back to it later.